src/System.Threading.Tasks.Extensions/tests/ManualResetValueTaskSource.cs

```diff
@@ -0,0 +1,171 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.ExceptionServices;
+using System.Threading.Tasks.Sources;
+
+namespace System.Threading.Tasks.Tests
+{
+    internal static class ManualResetValueTaskSource
+    {
+        public static ManualResetValueTaskSource<T> Completed<T>(T result, Exception error = null)
+        {
+            var vts = new ManualResetValueTaskSource<T>();
+            if (error != null)
+            {
+                vts.SetException(error);
+            }
+            else
+            {
+                vts.SetResult(result);
+            }
+            return vts;
+        }
+
+        public static ManualResetValueTaskSource<T> Delay<T>(int delayMs, T result, Exception error = null)
+        {
+            var vts = new ManualResetValueTaskSource<T>();
+            Task.Delay(delayMs).ContinueWith(_ =>
+            {
+                if (error != null)
+                {
+                    vts.SetException(error);
+                }
+                else
+                {
+                    vts.SetResult(result);
+                }
+            });
+            return vts;
+        }
+    }
+
+    internal sealed class ManualResetValueTaskSource<T> : IValueTaskSource<T>, IValueTaskSource
```

Comment:
This seems pretty close to a general purpose ValueTaskSource :)

Reply:
It is, but for a specific set of scenarios.  It wouldn't work, for example, in a situation where you need a different base type (e.g. SocketAsyncEventArgs), or where you effectively want it auto-reset once the last user is done with it (e.g. as in AsyncOperation), or other such things.

Reply:
When you say "need a different base type" I assume you mean "want to avoid an extra allocation by implementing IValueTaskSource on an existing class".  Right?

It's nice that we did this with internally in Sockets, but honestly, I question whether it really matters.  The whole point of both SAEA and IValueTaskSource is that they should be reused, ideally many, many times.  So the allocation cost is amortized and thus insignificant.

That's partly why I would really like to see some sort of general purpose class available.  Sure, if the extra-super-optimized case you would need a struct, and delegate the interface impl to it, etc, etc.  But in most real world cases, I'm very skeptical that the extra alloc matters -- especially when compared to what users are probably doing today, which is allocating a TaskCompletionSource *per operation*.

Reply:
> I assume you mean "want to avoid an extra allocation by implementing IValueTaskSource on an existing class". Right?

Yes.

> The whole point of both SAEA and IValueTaskSource is that they should be reused, ideally many, many times.

I would say for IValueTaskSource that they "can" be reused many, many times.  In channels, for example, you can have many producers and consumers all accessing a channel concurrently.  The most value for the least effort comes from caching a single IValueTaskSource instance and reusing that one, but when multiple producers or multiple consumers are all partying at the same time, only one of each is getting the reusable one, and the rest are all getting throwaways.  We could pool more, of course, but the management of that isn't free, and adds complexity.  So I do expect these to be used in non-reuse scenarios and I still want them to be as cheap as possible then.  I could of course have two different implementations for the two different cases, but that also adds complexity.

> That's partly why I would really like to see some sort of general purpose class available.

No one is arguing with that.  But even if you take the base task piece away, there are still differences in semantics that leads to needing to figure out which scenarios you care about and what knobs you want to expose.  This test ManualResetValueTaskSource wouldn't actually work as the implementation for any of the ones in the product because, well, they all automatically reset rather than manually resetting.